boards: cy8cproto_041p: Adds pwm support

Surya-Tumma opened the pull request:
Add Initial support for Infineon PSoC4 TCPWM driver update.

Surya-Tumma committed:
drivers: pwm: infineon: add support for PSoC4 family

Updates the Infineon TCPWM driver to support the PSoC4 series.
This includes SoC-specific register configurations, API calls,
and hardware instance mapping required for PSoC4 compatibility.

Signed-off-by: Surya Prakash T <suryat@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

Surya-Tumma committed:
drivers: pwm: infineon: refactor TCPWM driver for API abstraction

Refactors the Infineon TCPWM driver to abstract register access and
hardware control behind helper functions. This lays the groundwork for
supporting multiple SoC families (CAT1/PSoC4) by centralizing
platform-specific logic.

Changes:
- Replaced direct pointer access to `TCPWM_GRP_CNT_Type` with a more
  flexible `TCPWM_Type` base and instance index approach.
- Implemented static inline helper functions for common operations.
- Updated `ifx_tcpwm_pwm_set_cycles` to use the new abstracted helpers,
  making the main logic cleaner and more portable.
- Switched to using standard PDL APIs (e.g., `Cy_TCPWM_PWM_Enable`)
  instead of custom macros.

Signed-off-by: Surya Prakash T <suryat@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

Surya-Tumma committed:
dts: arm: infineon: psoc4: add PWM pinmux and TCPWM interrupts

Extends the PSoC 4100T Plus support by adding PWM pinmux definitions
and configuring TCPWM hardware interrupts.

Changes:
- Added comprehensive PWM and complementary PWM pinmux mappings in
  psoc4100tp.64-tqfp.dtsi for the 64-TQFP package.
- Defined interrupts (mapping to IRQs 16-21) and priorities for all
  six TCPWM instances (tcpwm0_0 to tcpwm0_5) in psoc4100tp.dtsi.

Signed-off-by: Surya Prakash T <suryat@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

Surya-Tumma committed:
samples: blinky_pwm: add overlay for cy8cproto_041tp

Adds the required Devicetree overlay to support the blinky_pwm
sample on the cy8cproto_041tp board, enabling PWM output
and pin configurations.

Signed-off-by: Surya Prakash T <suryat@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

jsbatch commented on drivers/pwm/pwm_infineon_tcpwm.c:
```diff
@@ -55,14 +60,19 @@ static int ifx_tcpwm_pwm_init(const struct device *dev)
+#else
+	status = IFX_TCPWM_PWM_Init(config->reg_base, &pwm_config);
+#endif
+
```
This PR needs to be refactored similar to the changes made in this PR:  https://github.com/zephyrproject-rtos/zephyr/pull/101310.   This will remove the need for these conditionals and make the driver much more readable.

jsbatch commented on drivers/pwm/pwm_infineon_tcpwm.c:
```diff
@@ -138,7 +150,37 @@ static int ifx_tcpwm_pwm_set_cycles(const struct device *dev, uint32_t channel,
+
+	Cy_TCPWM_PWM_Enable(config->reg_base, config->tcpwm_index);
+	Cy_TCPWM_TriggerStart(config->reg_base, BIT(config->tcpwm_index));
+#endif
```
The driver as implemented in this PR contains a large amount of duplicated code separated by #ifdefs, with only the IFX_ wrapper functions as a difference in many cases.  I'm counting more than 40 lines such lines of added code.  I consider that to be error prone and not very maintainable.  

The counter driver has been refactored in such a way that I see 6 instances of #ifdefs for PSOC4, each wrapping only a single line of code.  That is much more readable and maintainable, and should be the goal.  

Regarding testing, ping me when the changes are ready and I will run our hardware CI on the changes to validate other boards.

jsbatch commented on drivers/pwm/pwm_infineon_tcpwm.c:
```diff
@@ -138,7 +149,37 @@ static int ifx_tcpwm_pwm_set_cycles(const struct device *dev, uint32_t channel,
+	} else {
+		Cy_TCPWM_PWM_SetPeriod1(config->reg_base, config->tcpwm_index, period_cycles - 1);
+		Cy_TCPWM_PWM_SetCompare0(config->reg_base, config->tcpwm_index, pulse_cycles);
+		Cy_TCPWM_TriggerCaptureOrSwap(config->reg_base, config->tcpwm_index);
```
The CONFIG_SOC_FAMILY_INFINEON_CAT1 implementation is using buffered registers for the Period and Compare values.  I don't see that happening here.  Unless this hardware doesn't support buffering, we really should be using the buffered compare and period registers.  

Care needs to be taken when directly updating the compare and period registers.  If the period is shortened when the count is already higher than the new counter period length, you will miss the counter reset.  If the compare value is changed  below the current timer count, you will miss the compare for this PWM cycle.  When using the buffered registers, the changes take effect on the next PWM cycle instead.

jsbatch reviewed (changes requested):
This PR provides an overlay for the blinky_pwm sample.  I think it would be good to add overlays for the fade_led sample as well.  Additionally, I highly recommend creating an overlay and running the gwm_gpio_loopback test.  This test does a good job testing some of the common corner cases that can cause problems in a PWM implementation.

Surya-Tumma committed:
tests: drivers: pwm: add cy8cproto_041tp support for loopback test

Add board-specific configuration and Devicetree overlay for the
CY8CPROTO-041TP development board to enable the pwm_gpio_loopback test.
defining the loopback path in the overlay: PWM output on P3.1 and
GPIO input on P3.0.

Signed-off-by: Surya Prakash T <suryat@aerlync.com>
Signed-off-by: Sayooj K Karun <sayooj@aerlync.com>

Surya-Tumma commented on drivers/pwm/pwm_infineon_tcpwm.c:
```diff
@@ -100,54 +198,57 @@ static int ifx_tcpwm_pwm_set_cycles(const struct device *dev, uint32_t channel,
+	/* Validate cycles based on resolution */
+	ret = ifx_tcpwm_pwm_validate_cycles(config, period_cycles, pulse_cycles);
+	if (ret != 0) {
+		return ret;
```
Fixed.

jsbatch commented on tests/drivers/pwm/pwm_gpio_loopback/boards/cy8cproto_041tp.conf:
```diff
@@ -0,0 +1,2 @@
+CONFIG_GPIO=y
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=24000000
```
This points to an underlying clock misconfiguration.  We have already identified this issue internally, if you run the basic/blinky sample application, you will observe the LED blinks at the wrong rate as well.  I think it's fine to use this locally to sidestep the issue, but we shouldn't include it in the PR.  The underlying issue needs to be addressed instead.

Surya-Tumma replied:
@jsbatch Rebased, removed the conf file.

jsbatch commented on drivers/pwm/pwm_infineon_tcpwm.c:
```diff
@@ -71,24 +74,119 @@ static int ifx_tcpwm_pwm_init(const struct device *dev)
+#if defined(CONFIG_SOC_FAMILY_INFINEON_PSOC4)
+	Cy_TCPWM_TriggerStart(config->reg_base, BIT(config->tcpwm_index));
+#else
+	Cy_TCPWM_TriggerStart_Single(config->reg_base, BIT(config->tcpwm_index));
```
```suggestion
	Cy_TCPWM_TriggerStart_Single(config->reg_base, config->tcpwm_index);
```
This breaks non-psoc4 devices.

Surya-Tumma commented on drivers/pwm/pwm_infineon_tcpwm.c:
```diff
@@ -100,54 +198,58 @@ static int ifx_tcpwm_pwm_set_cycles(const struct device *dev, uint32_t channel,
+
+	/* If the PWM is not yet running, write the period and compare directly */
+	if (is_ifx_tcpwm_pwm_running(config)) {
+		ifx_tcpwm_pwm_set_initial_values(config, period_cycles, pulse_cycles);
```
changed.

Surya-Tumma commented on tests/drivers/pwm/pwm_gpio_loopback/boards/cy8cproto_041tp.overlay:
```diff
@@ -0,0 +1,49 @@
+		 * These pins must be physically connected for the test to pass.
+		 */
+		pwms = <&pwm0_1 0 PWM_MSEC(2) (PWM_POLARITY_NORMAL | PWM_IFX_TCPWM_OUTPUT_HIGHZ)>;
+		gpios = <&gpio_prt5 4 GPIO_ACTIVE_HIGH>;
```
Changed to P3.1 (PWM) output pin and P3.0 (GPIO) input pin.

JarmouniA reviewed (commented):
PR description should contain an actual description.

Surya-Tumma commented on tests/drivers/pwm/pwm_gpio_loopback/boards/cy8cproto_041tp.conf:
```diff
@@ -0,0 +1,2 @@
+CONFIG_GPIO=y
```
Moved CONFIG_GPIO to cy8cproto_041tp.conf because enabling it cy8cproto_041tp_defconfig in causes Twister failures.

jsbatch commented:
@Deepika-aerlync , @Surya-Tumma This PR needs to be rebased and merge conflicts resolved.

Deepika-aerlync commented:
> @Deepika-aerlync , @Surya-Tumma This PR needs to be rebased and merge conflicts resolved.

@jsbatch Rebased and resolved the merge conflicts.

teburd reviewed (commented):
The amount of ifdefs here mean we should make this a separate tcpwm driver. Please make this a distinct pwm driver without ifdef's for PSOC4 family. Use CMake and Kconfig to select which driver is built based on PSOC4 family Kconfig and otherwise use the existing driver.

The driver file can be named pwm_infineon_m0s8tcpwm.c

We are working on sorting out how best to do this still going forward but this at least should unblock the PR.
